File: go/internal/infra/memory/repositories.go
package memory

import (
	"context"
	"fmt"
	"sort"
	"time"

	"github.com/Yutarotaro/clean-architecture-billing/go/internal/domain"
	"github.com/Yutarotaro/clean-architecture-billing/go/internal/infra/storage"
	"github.com/Yutarotaro/clean-architecture-billing/go/internal/usecase"
)

// staging は 1 つの UnitOfWork の中で溜めた変更。
//
// commit されるまで Database には触れない。読み取りはまずここを見るので、同じ
// トランザクションの中では自分の書き込みが見える（read-your-writes）。
type staging struct {
	plans         map[domain.PlanID]domain.Plan
	subscriptions map[domain.SubscriptionID]*domain.Subscription
	invoices      map[domain.InvoiceID]*domain.Invoice
	versions      map[string]int
}

func newStaging() *staging {
	return &staging{
		plans:         map[domain.PlanID]domain.Plan{},
		subscriptions: map[domain.SubscriptionID]*domain.Subscription{},
		invoices:      map[domain.InvoiceID]*domain.Invoice{},
		versions:      map[string]int{},
	}
}

// versionTracker は読み出した集約のバージョンを覚えておく。
//
// ドメインオブジェクトに version フィールドを持たせない代わりに、リポジトリが
// 「この UnitOfWork の中で、この ID をバージョン幾つで読んだか」を記録する。
// ドメインモデルは楽観ロックの存在を最後まで知らない。
type versionTracker struct {
	entity   string
	versions map[string]int
}

func newVersionTracker(entity string) *versionTracker {
	return &versionTracker{entity: entity, versions: map[string]int{}}
}

func (t *versionTracker) remember(id string, version int) {
	t.versions[id] = version
}

func (t *versionTracker) expected(id string) (int, error) {
	version, ok := t.versions[id]
	if !ok {
		// 読まずに更新しようとしている。この状態では「誰も触っていないこと」を
		// 保証できないので、上書きせずに落とす。
		return 0, usecase.ConcurrencyConflict(t.entity, id)
	}
	return version, nil
}

func (t *versionTracker) bump(id string) int {
	t.versions[id]++
	return t.versions[id]
}

func (t *versionTracker) conflict(id string) error {
	return usecase.ConcurrencyConflict(t.entity, id)
}

type planRepo struct {
	db      *Database
	staging *staging
}

func (r *planRepo) Get(_ context.Context, id domain.PlanID) (*domain.Plan, error) {
	if plan, ok := r.staging.plans[id]; ok {
		return &plan, nil
	}
	r.db.mu.RLock()
	defer r.db.mu.RUnlock()
	plan, ok := r.db.plans[id]
	if !ok {
		return nil, nil
	}
	return &plan, nil
}

func (r *planRepo) ListAll(_ context.Context) ([]domain.Plan, error) {
	r.db.mu.RLock()
	merged := map[domain.PlanID]domain.Plan{}
	for id, plan := range r.db.plans {
		merged[id] = plan
	}
	r.db.mu.RUnlock()
	for id, plan := range r.staging.plans {
		merged[id] = plan
	}

	plans := make([]domain.Plan, 0, len(merged))
	for _, plan := range merged {
		plans = append(plans, plan)
	}
	sort.Slice(plans, func(i, j int) bool { return plans[i].ID < plans[j].ID })
	return plans, nil
}

func (r *planRepo) Add(_ context.Context, plan domain.Plan) error {
	r.staging.plans[plan.ID] = plan
	return nil
}

type subscriptionRepo struct {
	db       *Database
	staging  *staging
	versions *versionTracker
}

func (r *subscriptionRepo) Get(
	_ context.Context, id domain.SubscriptionID,
) (*domain.Subscription, error) {
	if staged, ok := r.staging.subscriptions[id]; ok {
		r.versions.remember(string(id), r.staging.versions[string(id)])
		return cloneSubscription(staged), nil
	}
	r.db.mu.RLock()
	defer r.db.mu.RUnlock()
	stored, ok := r.db.subscriptions[id]
	if !ok {
		return nil, nil
	}
	r.versions.remember(string(id), r.db.versions[string(id)])
	return cloneSubscription(stored), nil
}

func (r *subscriptionRepo) Add(_ context.Context, s *domain.Subscription) error {
	id := string(s.ID)
	if r.exists(s.ID) {
		return fmt.Errorf("%w: subscription %q already exists", storage.ErrDuplicate, id)
	}
	r.staging.subscriptions[s.ID] = cloneSubscription(s)
	r.staging.versions[id] = 1
	r.versions.remember(id, 1)
	return nil
}

func (r *subscriptionRepo) Save(_ context.Context, s *domain.Subscription) error {
	id := string(s.ID)
	if !r.exists(s.ID) {
		return fmt.Errorf("%w: subscription %q does not exist", storage.ErrUnknown, id)
	}
	expected, err := r.versions.expected(id)
	if err != nil {
		return err
	}
	// 楽観ロックの判定は、staging ではなくコミット済みの実体に対して行う。staging だけを
	// 見ていると、他のトランザクションが先に更新したことに永遠に気づけない。
	if baseline, ok := r.committedVersion(id); ok && baseline != expected {
		return r.versions.conflict(id)
	}
	r.staging.subscriptions[s.ID] = cloneSubscription(s)
	r.staging.versions[id] = r.versions.bump(id)
	return nil
}

func (r *subscriptionRepo) ListDue(
	_ context.Context, at time.Time, limit int,
) ([]*domain.Subscription, error) {
	found := r.selectSubscriptions(func(s *domain.Subscription) bool { return s.IsDue(at) })
	sort.Slice(found, func(i, j int) bool {
		return found[i].CurrentPeriod.End.Before(found[j].CurrentPeriod.End)
	})
	return r.take(found, limit), nil
}

func (r *subscriptionRepo) ListPastDue(
	_ context.Context, limit int,
) ([]*domain.Subscription, error) {
	found := r.selectSubscriptions(func(s *domain.Subscription) bool {
		return s.Status == domain.StatusPastDue
	})
	sort.Slice(found, func(i, j int) bool {
		left, right := found[i].PastDueSince, found[j].PastDueSince
		if left == nil || right == nil {
			return found[i].ID < found[j].ID
		}
		return left.Before(*right)
	})
	return r.take(found, limit), nil
}

func (r *subscriptionRepo) selectSubscriptions(
	keep func(*domain.Subscription) bool,
) []*domain.Subscription {
	r.db.mu.RLock()
	merged := map[domain.SubscriptionID]*domain.Subscription{}
	for id, s := range r.db.subscriptions {
		merged[id] = s
	}
	r.db.mu.RUnlock()
	for id, s := range r.staging.subscriptions {
		merged[id] = s
	}

	var found []*domain.Subscription
	for _, s := range merged {
		if keep(s) {
			found = append(found, cloneSubscription(s))
		}
	}
	return found
}

func (r *subscriptionRepo) take(all []*domain.Subscription, limit int) []*domain.Subscription {
	if limit > 0 && len(all) > limit {
		all = all[:limit]
	}
	for _, s := range all {
		id := string(s.ID)
		if version, ok := r.staging.versions[id]; ok {
			r.versions.remember(id, version)
			continue
		}
		if version, ok := r.committedVersion(id); ok {
			r.versions.remember(id, version)
		}
	}
	return all
}

func (r *subscriptionRepo) exists(id domain.SubscriptionID) bool {
	if _, ok := r.staging.subscriptions[id]; ok {
		return true
	}
	r.db.mu.RLock()
	defer r.db.mu.RUnlock()
	_, ok := r.db.subscriptions[id]
	return ok
}

func (r *subscriptionRepo) committedVersion(id string) (int, bool) {
	r.db.mu.RLock()
	defer r.db.mu.RUnlock()
	version, ok := r.db.versions[id]
	return version, ok
}

File: python/src/billing/infrastructure/persistence/memory/repositories.py
"""dict を使ったリポジトリ実装。

SQL 版・DynamoDB 版と同じ楽観ロックの意味論をここでも実装している。「テスト用だから
競合は起きない」と手を抜くと、契約テスト（tests/contract/）が 3 実装で同じ結果に
ならなくなり、インメモリで通ったテストの意味が消える。
"""

from __future__ import annotations

from copy import deepcopy
from datetime import datetime

from billing.domain.ids import CustomerId, InvoiceId, PlanId, SubscriptionId
from billing.domain.invoice import Invoice, InvoiceStatus
from billing.domain.plan import Plan
from billing.domain.subscription import Subscription, SubscriptionStatus
from billing.infrastructure.persistence.errors import DuplicateEntity, UnknownEntity
from billing.infrastructure.persistence.memory.database import MemoryDatabase, Staging
from billing.infrastructure.persistence.versioning import VersionTracker


class InMemoryPlanRepository:
    def __init__(self, db: MemoryDatabase, staging: Staging) -> None:
        self._db = db
        self._staging = staging

    def get(self, plan_id: PlanId) -> Plan | None:
        if plan_id in self._staging.plans:
            return self._staging.plans[plan_id]
        return self._db.plans.get(plan_id)

    def list_all(self) -> list[Plan]:
        merged = {**self._db.plans, **self._staging.plans}
        return sorted(merged.values(), key=lambda plan: plan.id)

    def add(self, plan: Plan) -> None:
        self._staging.plans[plan.id] = plan


class InMemorySubscriptionRepository:
    def __init__(self, db: MemoryDatabase, staging: Staging) -> None:
        self._db = db
        self._staging = staging
        self._versions = VersionTracker("subscription")

    def get(self, subscription_id: SubscriptionId) -> Subscription | None:
        staged = self._staging.subscriptions.get(subscription_id)
        if staged is not None:
            self._versions.remember(
                str(subscription_id), self._staging.versions[str(subscription_id)]
            )
            return deepcopy(staged)
        stored = self._db.subscriptions.get(subscription_id)
        if stored is None:
            return None
        self._versions.remember(str(subscription_id), self._db.versions[str(subscription_id)])
        # 複製して返す。参照をそのまま渡すと、呼び出し側がフィールドを書き換えた瞬間に、
        # commit していないのにデータベースの中身が変わる。
        return deepcopy(stored)

    def add(self, subscription: Subscription) -> None:
        entity_id = str(subscription.id)
        if self._exists(subscription.id):
            raise DuplicateEntity(f"subscription {entity_id!r} already exists")
        self._staging.subscriptions[subscription.id] = deepcopy(subscription)
        self._staging.versions[entity_id] = 1
        self._versions.remember(entity_id, 1)

    def save(self, subscription: Subscription) -> None:
        entity_id = str(subscription.id)
        if not self._exists(subscription.id):
            raise UnknownEntity(f"subscription {entity_id!r} does not exist")
        expected = self._versions.expected(entity_id)
        # 楽観ロックの判定は、staging ではなくコミット済みの実体に対して行う。
        # staging だけを見ていると、他のトランザクションが先に更新したことに気づけない。
        committed = self._db.versions.get(entity_id)
        if committed is not None and committed != expected:
            raise self._versions.conflict(entity_id)
        self._staging.subscriptions[subscription.id] = deepcopy(subscription)
        self._staging.versions[entity_id] = self._versions.bump(entity_id)

    def list_due(self, at: datetime, *, limit: int = 100) -> list[Subscription]:
        found = [s for s in self._merged().values() if s.is_due(at)]
        found.sort(key=lambda s: s.current_period.end)
        return self._take(found, limit)

    def list_past_due(self, *, limit: int = 100) -> list[Subscription]:
        found = [s for s in self._merged().values() if s.status is SubscriptionStatus.PAST_DUE]
        found.sort(key=lambda s: (s.past_due_since or s.current_period.start, s.id))
        return self._take(found, limit)

    def _merged(self) -> dict[SubscriptionId, Subscription]:
        return {**self._db.subscriptions, **self._staging.subscriptions}

    def _take(self, found: list[Subscription], limit: int) -> list[Subscription]:
        selected = found[:limit]
        for subscription in selected:
            entity_id = str(subscription.id)
            version = self._staging.versions.get(entity_id, self._db.versions.get(entity_id))
            if version is not None:
                self._versions.remember(entity_id, version)
        return [deepcopy(subscription) for subscription in selected]

    def _exists(self, subscription_id: SubscriptionId) -> bool:
        return (
            subscription_id in self._staging.subscriptions
            or subscription_id in self._db.subscriptions
        )


class InMemoryInvoiceRepository:
    def __init__(self, db: MemoryDatabase, staging: Staging) -> None:
        self._db = db
        self._staging = staging
        self._versions = VersionTracker("invoice")

    def get(self, invoice_id: InvoiceId) -> Invoice | None:
        staged = self._staging.invoices.get(invoice_id)
        if staged is not None:
            self._versions.remember(str(invoice_id), self._staging.versions[str(invoice_id)])
            return deepcopy(staged)
        stored = self._db.invoices.get(invoice_id)
        if stored is None:
            return None
        self._versions.remember(str(invoice_id), self._db.versions[str(invoice_id)])
        return deepcopy(stored)

    def add(self, invoice: Invoice) -> None:
        entity_id = str(invoice.id)
        if self._exists(invoice.id):
            raise DuplicateEntity(f"invoice {entity_id!r} already exists")
        if invoice.idempotency_key is not None and any(
            other.idempotency_key == invoice.idempotency_key for other in self._merged().values()
        ):
            # SQL の UNIQUE 制約、DynamoDB の条件付き書き込みに相当する検査。
            raise DuplicateEntity(f"idempotency key {invoice.idempotency_key!r} already used")
        self._staging.invoices[invoice.id] = deepcopy(invoice)
        self._staging.versions[entity_id] = 1
        self._versions.remember(entity_id, 1)

    def save(self, invoice: Invoice) -> None:
        entity_id = str(invoice.id)
        if not self._exists(invoice.id):
            raise UnknownEntity(f"invoice {entity_id!r} does not exist")
        expected = self._versions.expected(entity_id)
        committed = self._db.versions.get(entity_id)
        if committed is not None and committed != expected:
            raise self._versions.conflict(entity_id)
        self._staging.invoices[invoice.id] = deepcopy(invoice)
        self._staging.versions[entity_id] = self._versions.bump(entity_id)

    def find_by_idempotency_key(self, key: str) -> Invoice | None:
        for invoice in self._merged().values():
            if invoice.idempotency_key == key:
                self._remember(str(invoice.id))
                return deepcopy(invoice)
        return None

    def list_unsettled(self, *, issued_before: datetime, limit: int = 100) -> list[Invoice]:
        found = [
            invoice
            for invoice in self._merged().values()
            if invoice.status is InvoiceStatus.OPEN
            and invoice.issued_at is not None
            and invoice.issued_at <= issued_before
        ]
        found.sort(key=lambda invoice: (invoice.issued_at or datetime.min, invoice.id))
        return self._take(found, limit)

    def list_for_customer(self, customer_id: CustomerId) -> list[Invoice]:
        found = [i for i in self._merged().values() if i.customer_id == customer_id]
        found.sort(key=lambda i: (i.issued_at is None, i.issued_at or datetime.min, i.id))
        return self._take(found, len(found))

    def _merged(self) -> dict[InvoiceId, Invoice]:
        return {**self._db.invoices, **self._staging.invoices}

    def _take(self, found: list[Invoice], limit: int) -> list[Invoice]:
        selected = found[:limit]
        for invoice in selected:
            self._remember(str(invoice.id))
        return [deepcopy(invoice) for invoice in selected]

    def _remember(self, entity_id: str) -> None:
        version = self._staging.versions.get(entity_id, self._db.versions.get(entity_id))
        if version is not None:
            self._versions.remember(entity_id, version)

    def _exists(self, invoice_id: InvoiceId) -> bool:
        return invoice_id in self._staging.invoices or invoice_id in self._db.invoices
